[DOCS] Adds frequent items agg docs (#86037)
elastic/elasticsearch: docs/reference/aggregations/bucket/frequent-items-aggregation.asciidoc

intensive. This parameter has an effect on memory and runtime.

[discrete]
[[frequent-items-minimum-set-size]]
==== Minimum set size

The minimum set size is the minimum number of item the set needs to contain. A 
value of 1 returns the frequency of single items. The higher the minimum set 
size the less items are returned. For example, the item set `apple, banana, 
orange` is only returned if the minimum set size is 3 or lower.

[discrete]
[[frequent-items-size]]
==== Size

This parameter defines the maximum number of item sets to return. The result 
contains top-k item sets; the item sets with the highest support values. This 
parameter has a significant effect on memory and runtime.

[discrete]
[[frequent-items-example]]
==== Example

In this example, we use the e-commerce {kib} sample data set. The goal is to 
find out from what product categories customers frequently purchase products 
together based on transaction data. We are interested in sets with two items, 
and want to see the first three frequent item setes with the highest support. 

[source,console]
-------------------------------------------------

> GET kibana_sample_data_ecommerce /_search 
> {
>  "size": 0,
> 	"aggs": {
> 		"my_agg": {
> 			"frequent_items": {
> 				"minimum_set_size": 2,
> 				"fields": [ 
>           {
>             "field": "category.keyword"
>           }
> 				],
> 				"size": 3
> 			}
> 		}
> 	}
> }

Author: ```suggestion
GET kibana_sample_data_ecommerce /_search 
{
 "size": 0,
   "aggs": {
     "my_agg": {
       "frequent_items": {
         "minimum_set_size": 2,
	 "fields": [ 
            {
              "field": "category.keyword"
            }
	 ],
	 "size": 3
       }
     }
   }
}
```

[[search-aggregations-bucket-frequent-items-aggregation]]
=== Frequent items aggregation
++++
<titleabbrev>Frequent items</titleabbrev>
++++

A bucket aggregation which finds frequent item sets. Finding frequent item sets 

> is a form of association rules mining that identifies items that are often occur 

Reviewer: ```suggestion
is a form of association rules mining that identifies items that often occur 
```

together. Items that are frequently purchased together or log events that tend 
to co-occur are examples of frequent item sets. Finding frequent item sets helps 
to discover relationships between different data points (items) in a data set 

> and to create association rules based on those relationships.

Reviewer: > and to create association rules based on those relationships.

"association rules" isn't a concept I'm familiar with and didn't seem to pop up anywhere else in a quick docs search. It might be a good idea to give a quick explanation or link elsewhere for more context.

> 

Reviewer 2: We should mention that `frequent_items` reports "closed itemsets":

"A frequent item set is called closed if no superset has the same support."

That's very technical, you maybe find a better way to formulate it. It simply means that if you have candidates like:

cat, dog, crocodile -> 0.2
cat, dog, crocodile, river -> 0.2

only cat, dog, crocodile, river gets returned, the subset cat, dog, crocodile gets skipped. On the other side for the sets: 

cat, doc -> 0.4
cat, dog, crocodile -> 0.2
cat, dog, crocodile, river -> 0.2

Both cat, doc -> 0.4 and cat, dog, crocodile, river -> 0.2 are returned, because cat, doc has a higher/different support.

==== Syntax

A `frequent_items` aggregation looks like this in isolation:

[source,js]
--------------------------------------------------
"frequent_items": {
  "minimum_set_size": 3,
  "fields": [
    {"field": "my_field_1"},
    {"field": "my_field_2"}
  ]
}
--------------------------------------------------
// NOTCONSOLE

.`frequent_items` Parameters
|===
|Parameter Name |Description |Required |Default Value

> |`field` |(array) Fields to analyze. | Required |

Reviewer 2: `field` -> `fields`

> |`minimum_set_size` | (integer) The minimum size of one item set. | Optional | `1`

Reviewer: If you choose to add one link, we should add both:
```suggestion
|`minimum_set_size` | (integer) The <<frequent-items-minimum-set-size,minimum size>> of one item set. | Optional | `1`
```

> |`minimum_support` | (integer) The minimum support of one item set. | Optional | `0.1`

Reviewer: I read this and thought "I don't know what minimum support means", so it might be nice to link to the details here
```suggestion
|`minimum_support` | (integer) The <<frequent-items-minimum-support,minimum support>> of one item set. | Optional | `0.1`
```

> |`size` | (integer) The number of top item sets to return. | Optional |

Reviewer 2: Default: 10

|===

[discrete]
[[frequent-items-minimum-support]]

> ==== Minimum support

Reviewer: > ==== Minimum support

Not mandatory, but it might look nice to order these subsections the same way they're ordered in the table (i.e. size then support).

Support represents the frequency of an item set normalized between 0 
and 1. It is calculated by dividing the number of documents containing the item 
set by the total number of documents.

> For example, if a given item set is contained by five documents and the total 

Reviewer: > For example, if a given item set is contained by five documents and the total 

Is "contained by" correct? I would have guessed that an item set consisted of five documents instead.

Author: Yes, I think `contained by` is correct. A document is for example a transaction list that contains purchased items [banana, apple, beer, diapers, eggs]. And you have for example 20 documents that are basically transaction lists from which 5 documents might contain `banana, apple, orange`.

number of documents is 20, then the support of the item set is 5/20 = 0.25. 

> Therefore, this set is returned only if the minimum support is 0.25 or lower. As 

Reviewer: > Therefore, this set is returned only if the minimum support is 0.25 or lower. As 

I don't think I understand this section. It sounds like you'd only set this property if you already know the ratio of documents that contain your "frequent items" (how would you know that?). And by setting the property to this ratio, you're helping it quickly ignore/prune more of the items from its results. Is that the gist?

Author: You set the property to return only the item sets with the minimum support you are interested in. You don't know which item sets will be returned, but you know that – for example – every item sets with a minimum support that less than 0.5 is not interesting for you, so you set minimum support of 0.5 to get the items you are interested in.

a lower minimum support prunes more items, the calculation is less resource 
intensive. The `minimum_support` parameter has an effect on the required memory 
and the runtime of the aggregation.

[discrete]
[[frequent-items-minimum-set-size]]
==== Minimum set size

> The minimum set size is the minimum number of item the set needs to contain. A 

Reviewer: ```suggestion
The minimum set size is the minimum number of items the set needs to contain. A 
```

value of 1 returns the frequency of single items. The higher the minimum set 

> size the less items are returned. For example, the item set `cat, dog, 

Reviewer: > For example, the item set `cat, dog, 

I found this example confusing since those animal terms aren't referenced in any other examples here and the larger context is therefore unclear. Can we use an example that fits better with the other examples or omit it?

crocodile` is only returned if the minimum set size is 3 or lower.

[discrete]
[[frequent-items-size]]
==== Size

This parameter defines the maximum number of item sets to return. The result 
contains top-k item sets; the item sets with the highest support values. This 
parameter has a significant effect on the required memory and the runtime of the 
aggregation.

[discrete]
[[frequent-items-example]]
==== Example

In the following example, we use the e-commerce {kib} sample data set. The goal 
is to find out from what product categories the customers purchase products 
frequently together based on transaction data. We are interested in sets with 

> two items, and want to see the first three frequent item setes with the highest 

Reviewer: ```suggestion
two items, and want to see the first three frequent item sets with the highest 
```

> two items, and want to see the first three frequent item setes with the highest 

Reviewer: > two items, and want to see the first three frequent item setes with the highest 

Does this find sets with two *or more* items? Or only exactly two items?

support.

[source,console]
-------------------------------------------------
GET kibana_sample_data_ecommerce /_search 
{
  "size": 0,
  "aggs": {
    "my_agg": {
      "frequent_items": {
        "minimum_set_size": 2,
        "fields": [
          {
            "field": "category.keyword"

>           }

Reviewer 2: I would really like an example with at least 2 fields, can't we use one of the geoip fields in addition?

Author: Sure, I added it via [2a7e3fd](https://github.com/elastic/elasticsearch/pull/86037/commits/2a7e3fdd7377d14483e4a794c56b69eb78b04353).

>             {
>               "field": "category.keyword"
>             }

Author: ```suggestion
              {
              "field": "category.keyword"
              }
```

> 	       "fields": [ 
>               {
>               "field": "category.keyword"
>               }

Author: ```suggestion
	       "fields": [ 
                  {
                    "field": "category.keyword"
                  }
```

        ],
        "size": 3
      }
    }
  }
}
-------------------------------------------------
// TEST[skip:setup kibana sample data]

The API returns a response similar to the following one:

[source,console-result]
-------------------------------------------------
(...)
"aggregations" : {
    "my_agg" : {
      "buckets" : [
        {
          "key" : {
            "category.keyword" : [
              "Women's Clothing",
              "Women's Shoes"
            ]
          },
          "doc_count" : 738,
          "support" : 0.15786096256684493
        },
        {
          "key" : {
            "category.keyword" : [
              "Men's Clothing",
              "Men's Shoes"
            ]
          },
          "doc_count" : 695,
          "support" : 0.14866310160427806
        },
        {
          "key" : {
            "category.keyword" : [
              "Women's Clothing",
              "Women's Accessories"
            ]
          },
          "doc_count" : 449,
          "support" : 0.0960427807486631
        }
      ],
    (...) 
  }
}
-------------------------------------------------
// TEST[skip:setup kibana sample data]

> The response shows that the categories customers purcashe from most frequently 

Reviewer 2: purcashe -> purchase

[[search-aggregations-bucket-frequent-items-aggregation]]
=== Frequent items aggregation
++++
<titleabbrev>Frequent items</titleabbrev>
++++

A bucket aggregation which finds frequent item sets. Finding frequent item sets 
is a form of association rules mining that identifies items that often occur 
together. Items that are frequently purchased together or log events that tend 
to co-occur are examples of frequent item sets. Finding frequent item sets helps 
to discover relationships between different data points (items).

The aggregation reports closed item sets. A frequent item set is called closed 

> if no superset has the same support. For example, we have the two following 

Reviewer: Since we haven't explain "support" yet, perhaps this is simpler to understand:
```suggestion
if no superset exists with the same ratio of documents (also known as its <<frequent-items-minimum-support,support value>>). For example, we have the two following 
```

> candidates for a frequent item set **with the same support value**:

Reviewer: IMO the bold stands out too much:
```suggestion
candidates for a frequent item set, which have the same support value:
```

1. `apple, orange, banana`
2. `apple, orange, banana, tomato`.
Only the second item set (`apple, orange, banana, tomato`) is returned, and the 
first set – wich is a subset of the second one – is skipped. Both item sets 

> might be returned if their support values are different. Refer to 

Reviewer: > Refer to ...

If you accept my earlier suggestion about linking earlier in the paragraph, this sentence can be removed.

[source,js]
--------------------------------------------------
"frequent_items": {
  "minimum_set_size": 3,
  "fields": [
    {"field": "my_field_1"},
    {"field": "my_field_2"}
  ]
}
--------------------------------------------------
// NOTCONSOLE

.`frequent_items` Parameters
|===
|Parameter Name |Description |Required |Default Value
|`fields` |(array) Fields to analyze. | Required |
|`minimum_set_size` | (integer) The <<frequent-items-minimum-set-size,minimum size>> of one item set. | Optional | `1`
|`minimum_support` | (integer) The <<frequent-items-minimum-support,minimum support>> of one item set. | Optional | `0.1`
|`size` | (integer) The number of top item sets to return. | Optional | `10`
|===

[discrete]
[[frequent-items-minimum-set-size]]
==== Minimum set size

The minimum set size is the minimum number of items the set needs to contain. A 
value of 1 returns the frequency of single items. The higher the minimum set 
size the less items are returned. For example, the item set `orange, banana, 
apple` is only returned if the minimum set size is 3 or lower.

[discrete]
[[frequent-items-minimum-support]]
==== Minimum support

> Support represents the frequency of an item set normalized between 0 

Reviewer: Related to https://github.com/elastic/elasticsearch/pull/86037#discussion_r857348240, this is the way I had to wrap around the concept to make sense in my own mind. Hope it's useful:
```suggestion
The minimum support value is the ratio of documents that an item set must exist in to be considered "frequent". In particular, it is a normalized value between 0
```

and 1. It is calculated by dividing the number of documents containing the item 
set by the total number of documents.

For example, if a given item set is contained by five documents and the total 
number of documents is 20, then the support of the item set is 5/20 = 0.25. 
Therefore, this set is returned only if the minimum support is 0.25 or lower. 
As a higher minimum support prunes more items, the calculation is less resource 
intensive. The `minimum_support` parameter has an effect on the required memory 
and the runtime of the aggregation.

[discrete]
[[frequent-items-size]]
==== Size

This parameter defines the maximum number of item sets to return. The result 
contains top-k item sets; the item sets with the highest support values. This 
parameter has a significant effect on the required memory and the runtime of the 
aggregation.

[discrete]
[[frequent-items-example]]
==== Example

In the following example, we use the e-commerce {kib} sample data set. The goal 

> is to find out based on transaction data that (1.) from what product categories 

Reviewer: ```suggestion
is to find out based on transaction data (1.) from what product categories 
```

the customers purchase products frequently together and (2.) from which cities 

> they purchase those. We are interested in sets with three or more items, and 

Reviewer: ```suggestion
they make those purchases. We are interested in sets with three or more items, and 
```

              "Cairo"
            ]
          },
          "doc_count" : 123,
          "support" : 0.026310160427807486
        }
      ],
    (...) 
  }
}
-------------------------------------------------
// TEST[skip:setup kibana sample data]

The response shows that the categories customers purchase from most frequently 
together are `Women's Clothing` and `Women's Shoes` and customers from New York 
tend to buy items from these categories frequently togeher. In other words, 
customers who buy products labelled Women's Clothing more likely buy products 
also from the Women's Shoes category and customers from New York most likely buy 
products from these categories together. The item set with the second highest 
support is `Women's Clothing` and `Women's Accessories` with customers mostly 
from New York. Finally, the item set with the third highest support is 
`Men's Clothing` and `Men's Shoes` with customers mostly from Cairo.

The frequent items aggregation enables you to bucket numeric values by using 
<<runtime,runtime fields>>. The next example demonstrates how to use a script to 
add a runtime field to your documents that called `price_range` which is 

> calculated from the texful total price of the individual transactions. The 

Reviewer 2: texful->taxful
